REGRESSION (iOS 26.1): videoElement.audioTrack incorrectly changes "description" kind to main when the system settings prefer it

roberto-apple opened the pull request:
#### 3bb5190507aeecb273771abc40745229c08cb3b9
<pre>
REGRESSION (iOS 26.1): videoElement.audioTrack incorrectly changes &quot;description&quot; kind to main when the system settings prefer it
<a href="https://rdar.apple.com/165229625">rdar://165229625</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302933">https://bugs.webkit.org/show_bug.cgi?id=302933</a>

Reviewed by Eric Carlson.

When the Audio Descriptions accessibility setting is enabled, HLS audio
description tracks incorrectly report kind=&quot;main&quot; instead of
kind=&quot;description&quot;. The issue occurs when AVFoundation strips
accessibility characteristics from AVAssetTrack objects while preserving
them in associated MediaSelectionOption objects.

Add a fix to check MediaSelectionOption for accessibility
characteristics when they are missing from AVAssetTrack, ensuring that
audio description tracks receive correct classification.

Add a layout test that simulates the system condition where
MediaAccessibility return preferred characteristics, reproducing the
scenario where AVAssetTrack loses accessiblity characteristics while
MediaSelectionOption preserves them.

Test: http/tests/media/hls-audio-description-track-kind.html

* LayoutTests/http/tests/media/hls/hls-audio-description-track-kind-expected.txt: Added.
* LayoutTests/http/tests/media/hls/hls-audio-description-track-kind.html: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/ad.fmp4.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/ad.m4s: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/audio.fmp4.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/audio.init: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/audio.m4s: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/main.fmp4.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/main.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/video.init: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/video.m4s: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/page/CaptionUserPreferences.cpp:
(WebCore::CaptionUserPreferences::preferredAudioCharacteristics const):
* Source/WebCore/page/CaptionUserPreferences.h:
(WebCore::CaptionUserPreferences::setPreferredAudioCharacteristicsForTesting):
(WebCore::CaptionUserPreferences::preferredAudioCharacteristicsForTesting const):
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::audioKind const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setPreferredAudioCharacteristicsForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/306249@main">https://commits.webkit.org/306249@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2aec446c70745a706541c0c1164587891b3888e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140716 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/13098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/2260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/149067 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/93803 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/a2cd1564-41f5-4126-b13a-43e8df9fb162) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/142589 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13810 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/13252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/107910 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/78334 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/143667 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/10674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/125976 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/88812 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/b0d3e330-90c5-491a-ae2d-bed42c59ca58) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/10283 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/7840 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/9149 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/119529 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/1985 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/151679 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/12786 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/2150 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/116206 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12801 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/11127 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/116544 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29647 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/12527 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/122607 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/67925 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12828 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/2046 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/12568 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/76528 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/12767 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/12612 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/http/tests/media/hls-audio-description-track-kind.html:
```diff
@@ -0,0 +1,66 @@
+        testExpected("video.audioTracks.length", 2);
+
+        adTrack = null;
+        mainTrack = null;
```
Nit: this is unnecessary as they are initialized to `null`

eric-carlson commented on LayoutTests/TestExpectations:
```diff
@@ -315,6 +315,9 @@ webrtc/peerConnection-rvfc.html [ Skip ]
 
+# Test is a timeout on mac-AS-debug-wk2
+[ Debug ] http/tests/media/hls-audio-description-track-kind.html [ Skip ]
+
```
Please fix this before landing

roberto-apple committed:
REGRESSION (iOS 26.1): videoElement.audioTrack incorrectly changes "description" kind to main when the system settings prefer it
rdar://165229625
https://bugs.webkit.org/show_bug.cgi?id=302933

Reviewed by Eric Carlson.

When the Audio Descriptions accessibility setting is enabled, HLS audio
description tracks incorrectly report kind="main" instead of
kind="description". The issue occurs when AVFoundation strips
accessibility characteristics from AVAssetTrack objects while preserving
them in associated MediaSelectionOption objects.

Add a fix to check MediaSelectionOption for accessibility
characteristics when they are missing from AVAssetTrack, ensuring that
audio description tracks receive correct classification.

Add a layout test that simulates the system condition where
MediaAccessibility return preferred characteristics, reproducing the
scenario where AVAssetTrack loses accessiblity characteristics while
MediaSelectionOption preserves them.

Test: http/tests/media/hls-audio-description-track-kind.html

* LayoutTests/http/tests/media/hls/hls-audio-description-track-kind-expected.txt: Added.
* LayoutTests/http/tests/media/hls/hls-audio-description-track-kind.html: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/ad.fmp4.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/ad.m4s: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/audio.fmp4.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/audio.init: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/audio.m4s: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/main.fmp4.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/main.m3u8: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/video.init: Added.
* LayoutTests/http/tests/media/resources/hls/audio-description/video.m4s: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/page/CaptionUserPreferences.cpp:
(WebCore::CaptionUserPreferences::preferredAudioCharacteristics const):
* Source/WebCore/page/CaptionUserPreferences.h:
(WebCore::CaptionUserPreferences::setPreferredAudioCharacteristicsForTesting):
(WebCore::CaptionUserPreferences::preferredAudioCharacteristicsForTesting const):
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::audioKind const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setPreferredAudioCharacteristicsForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/306249@main
